src/Database/LSMTree/Internal.hs

```diff
@@ -541,13 +548,37 @@ updates ::
   -> m ()
 updates es th = withOpenTable th $ \thEnv -> do
     -- A placeholder implementation that is sufficient to pass the tests, but
-    -- keeps all entries in memory.
-    -- TODO: flush write buffer when full
+    -- simply writes small runs to disk without merging.
     -- TODO: merge runs when level becomes full
-    modifyMVar_ (tableWriteBuffer thEnv) $ \wb -> do
-      return $ foldl' (flip (uncurry (WB.addEntry resolve))) wb es
-  where
-    resolve = resolveMupsert (tableConfig th)
+    modifyMVar_ (tableContent thEnv) $ \tableContent -> do
+      let !wb = WB.addEntries (resolveMupsert (tableConfig th)) es $
+                  tableWriteBuffer tableContent
+      if WB.sizeInBytes wb <= fromIntegral (confWriteBufferAlloc (tableConfig th))
+        then return tableContent { tableWriteBuffer = wb }
+        else flushWriteBuffer thEnv (tableLevels tableContent) wb
+
+{-# INLINE flushWriteBuffer #-}
+flushWriteBuffer ::
+     m ~ IO
+  => TableHandleEnv m h
+  -> Levels (Handle h)
+  -> WriteBuffer
+  -> m (TableContent h)
+flushWriteBuffer thEnv levels wb = do
+    n <- incrUniqCounter (tablesSessionUniqCounter thEnv)
+    let runPaths = Paths.runPath (tableSessionRoot thEnv) n
+    run <- Run.fromWriteBuffer (tableHasFS thEnv) runPaths wb
+    let levels' = addRunToLevels run levels
+    return TableContent {
+        tableWriteBuffer = WB.empty
+      , tableLevels = levels'
+      , tableCache = mkLevelsCache levels'
+      }
```

Comment:
Maybe this should only partially flush the write buffer: a batch of updates might have overfilled the write buffer by a lot

Reply:
Wouldn't matter if we go by key count only.

Reply:
Oh I see. Perhaps it'd be simpler to split a batch of inserts so that we never over-fill the write buffer.

Or even simpler, if the batch would take us over the keys limit, flush the buffer as-is (so it's not 100% full) and put the new batch into the new (empty) buffer. That works except when the buffer is hardly full but the batch is so large it'd take us over the limit. In that case we'd need to split batches of inserts.